src/reader.rs

```diff
@@ -0,0 +1,133 @@
+        let entries_offset = self.cursor + HEADER_LEN as u64;
+        let entries_len = offset as usize - HEADER_LEN;
+
+        // skip entries
+        self.cursor += offset;
+
+        // read footer & recover
+        needed_buffer_size = len - offset as usize;
+        trace!("LogBatch::recovery_size = {}", needed_buffer_size);
+        if self.buffer_remain_size() < needed_buffer_size {
+            self.extend_buffer(needed_buffer_size, HEADER_LEN)?;
+        }
+        trace!(
+            "recover LogBatch offset:{} len:{}",
+            self.cursor,
+            needed_buffer_size
+        );
+        let batch = LogBatch::<M>::from_footer(
+            &mut self.buffer_slice_from_cursor(needed_buffer_size),
+            entries_offset,
+            entries_len,
+            compression_type,
+        )?;
+        self.cursor += needed_buffer_size as u64;
+        Ok(Some(batch))
+    }
+
+    pub fn buffer_remain_size(&self) -> usize {
+        if self.cursor as usize > self.offset as usize + self.len {
+            0
+        } else {
+            self.offset as usize + self.len - self.cursor as usize
+        }
+    }
+
+    fn buffer_slice_from_cursor(&self, len: usize) -> &[u8] {
+        trace!(
+            "::slice buffer offset:{} cursor:{}",
+            self.offset,
+            self.cursor
+        );
+        assert!(self.offset <= self.cursor);
+        assert!(self.offset as usize + self.len >= self.cursor as usize + len);
+        let start = (self.cursor - self.offset) as usize;
+        let end = start + len;
+        &self.buffer[start..end]
+    }
+
+    fn extend_buffer(&mut self, needed_buffer_size: usize, tail_hint: usize) -> Result<()> {
+        self.fread(self.cursor, needed_buffer_size + tail_hint)?;
+        if (self.offset as usize + self.len - self.cursor as usize) < needed_buffer_size {
+            return Err(Error::Corruption(format!(
+                "unexpected eof at {}",
+                self.offset + self.len as u64
+            )));
+        }
+        Ok(())
+    }
+
+    fn fread(&mut self, offset: u64, len: usize) -> Result<()> {
+        trace!("called fread at {}:{}", offset, len);
+        let offset = std::cmp::max(self.offset + self.len as u64, offset);
+        let mut len = std::cmp::max(len, self.read_block_size);
+        if offset as usize + len > self.fsize {
+            len -= offset as usize + len - self.fsize;
+        }
+
+        if offset == self.offset + self.len as u64 {
+            trace!("::extend buffer {}:{}", offset, len);
+            self.buffer.extend(self.fd.read(offset as i64, len)?);
+        } else {
+            trace!("::replace buffer {}:{}", offset, len);
+            self.buffer = self.fd.read(offset as i64, len)?;
+            self.offset = offset;
+        }
+        self.len = self.buffer.len();
+        trace!("new buffer {}:{}", self.offset, self.len);
+        Ok(())
+    }
```

Comment:
I don't think that's a good idea. These four function has different purpose, and some of them are reused multiple times. I should add some comments for docs for these function to make them more understandable.

Reply:
They are reused only because you leak out the details of reading data to buffer. DRY won't be violated If you merge them into `peek`, instead the code is much cleaner because I can see how all the internal states are mutated in one single function.

Reply:
I get you. Let me fix it.